cmd/query/app/server.go

```diff
@@ -297,52 +295,53 @@ func (s *Server) Start() error {
+		s.telset.Logger.Info("Starting HTTP server", zap.Int("port", httpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
 		var err error
 		if s.queryOptions.TLSHTTP.Enabled {
 			err = s.httpServer.ServeTLS(s.httpConn, "", "")
 		} else {
 			err = s.httpServer.Serve(s.httpConn)
 		}
 		if err != nil && !errors.Is(err, http.ErrServerClosed) && !errors.Is(err, cmux.ErrListenerClosed) && !errors.Is(err, cmux.ErrServerClosed) {
-			s.logger.Error("Could not start HTTP server", zap.Error(err))
+			s.telset.Logger.Error("Could not start HTTP server", zap.Error(err))
 		}
-		s.logger.Info("HTTP server stopped", zap.Int("port", httpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
-		s.healthCheck.Set(healthcheck.Unavailable)
+		s.telset.Logger.Info("HTTP server stopped", zap.Int("port", httpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
 		s.bgFinished.Done()
 	}()
 
 	// Start GRPC server concurrently
 	s.bgFinished.Add(1)
 	go func() {
-		s.logger.Info("Starting GRPC server", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
+		s.telset.Logger.Info("Starting GRPC server", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
 
 		err := s.grpcServer.Serve(s.grpcConn)
 		if err != nil && !errors.Is(err, cmux.ErrListenerClosed) && !errors.Is(err, cmux.ErrServerClosed) {
-			s.logger.Error("Could not start GRPC server", zap.Error(err))
+			s.telset.Logger.Error("Could not start GRPC server", zap.Error(err))
 		}
-		s.logger.Info("GRPC server stopped", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
-		s.healthCheck.Set(healthcheck.Unavailable)
+		s.telset.Logger.Info("GRPC server stopped", zap.Int("port", grpcPort), zap.String("addr", s.queryOptions.GRPCHostPort))
 		s.bgFinished.Done()
 	}()
 
 	// Start cmux server concurrently.
 	if !s.separatePorts {
 		s.bgFinished.Add(1)
 		go func() {
-			s.logger.Info("Starting CMUX server", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
+			s.telset.Logger.Info("Starting CMUX server", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
 
 			err := cmuxServer.Serve()
 			// TODO: find a way to avoid string comparison. Even though cmux has ErrServerClosed, it's not returned here.
 			if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
-				s.logger.Error("Could not start multiplexed server", zap.Error(err))
+				s.telset.Logger.Error("Could not start multiplexed server", zap.Error(err))
 			}
-			s.logger.Info("CMUX server stopped", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
-			s.healthCheck.Set(healthcheck.Unavailable)
+			s.telset.Logger.Info("CMUX server stopped", zap.Int("port", tcpPort), zap.String("addr", s.queryOptions.HTTPHostPort))
 			s.bgFinished.Done()
 		}()
 	}
-
+	if err != nil {
```

Comment:
The `err` here has nothing to do with the state of the background goroutines. You can have `err == nil` yet the servers will fail to start. It's incorrect to remove `s.healthCheck.Set(healthcheck.Unavailable)` and not use any replacement.

I checked OTLP Receiver and the only thing it currently does is report FatalError. I suggest we do the same for now, i.e. not bother with Starting / OK / Stopped states and only report
  * FatalError when server failed to start
  * Stopped when exiting goroutine

Both of those will be technically incorrect due to a different level of granularity of status reporting, but I don't see what we can do about that right now. I opened a ticket upstream https://github.com/open-telemetry/opentelemetry-collector/issues/10549

Reply:
Ok, so should I revert the code back to report a Fatal Error for each server?

Reply:
yes, but we never fatal reported, only Unavailable. I suggest we do both: Fatal, and Stopped.

Reply:
Done. Stopped can be reported when Stop() function is called. I have added the FatalEvents in case an err is created in initialising each server